src/overlay/Peer.cpp

```diff
@@ -388,6 +424,10 @@ Peer::shutdown()
 void
 Peer::connectHandler(asio::error_code const& error)
 {
+    // Thread-safe
+    std::lock_guard<std::recursive_mutex> guard(
+        mOverlayManager.getOverlayManagerMutex());
+
```

Comment:
Why does this function need the `OverlayManager`'s mutex?

Reply:
A few things are happening in this function: we're updating peer's state, or triggering peer drop. Both actions touch data that's read and written by main and background threads. For example, dropping a connection can be triggered from background (like if reading bytes from the wire failed) or from main thread (if we received a valid message, but a higher level processor decided that peer is not following the protocol)

Reply:
That makes sense, thanks!